lib/puppet/parser/compiler.rb

```diff
@@ -471,7 +471,7 @@ def ensure_classes_without_parameters(scope, hostclasses)
     newly_included = []
     hostclasses.each do |klass|
       class_scope = scope.class_scope(klass)
-      if class_scope
+      if class_scope && !class_scope.is_nodescope?
```

Comment:
Definitely could use a comment, because the behavior is super counterintuitive. IIUC this happens because `scope.class_scope()` can return _either_ a class scope _or_ a node scope. So when checking whether the class has already been included, we need to only check whether it's a _class_ that was already included?

What happens if we've already included a node with the same name _and_ a class with the same name? Could `scope.class_scope()` still return the node scope in that case? If so, we may get back the node scope here, think that means this isn't a conflict (even though we really did already include the class), and call `klass.ensure_in_catalog(scope)`.

That case should still be safe, because `ensure_in_catalog` will _properly_ check whether the resource is already in the catalog (not conflicting with a node of the same name) and return it. Then because this resource is in the `newly_included` array, the caller of this method will call `evaluate` on it. As the resource presumably will have already been evaluated, that will do nothing.

The other question is whether we might inadvertently look up the wrong scope _elsewhere_, since this fix will now allow us to have two "class" scopes with the same name. Since this code seems to be the only place that method is really used, I think that should also be safe.

Reply:
Great feedback, thanks! I'll update this.

Reply:
> So when checking whether the class has already been included, we need to only check whether it's a class that was already included?

Yep, thanks for asking these questions! This is a twisty passage:

1. Luke added a check in 0ff7827d4d7f42fe59c10af35266f197e83b2b17 to prevent a node and class having the same name.
2. Markus moved the `Compiler#class_set` method to `Scope` in 8bafc37a532b7bef541186e7bb719f50c0eda600
3. Luke removed the check as part of "Removing obsolete nodescope concept" in commit 938fbe914a811b10146f817368883df8180ef224
4. Later Brice modified Puppet to call [`class_set` with the node scope](https://github.com/puppetlabs/puppet/commit/153d7cd69b8b5216695206cbc235dd89a65dec76#diff-0a2ef5f959d5747b9f7414d936653e60d6f6808db0332a5662f281335d9c48d0R222). This was done to support node inheritance as described in [redmine 3665](http://web.archive.org/web/20151212201917/https://projects.puppetlabs.com/issues/3665). But it meant that if a node and class had the same name, then `Scope#class_scopes` would overwrite the former with the latter.

Reply:
So based on the above, I believe the following to be true:

> What happens if we've already included a node with the same name and a class with the same name? Could scope.class_scope() still return the node scope in that case?

I don't think this can happen, because the first time `Scope#class_set` is called with the class scope, it will overwrite the node scope. And [due to the way `class_set` walks its parent scopes](https://github.com/puppetlabs/puppet/blob/fc8dd4771cf736e44b20d4c017d79d574d223b70/lib/puppet/parser/scope.rb#L406-L410), we always set the node or class scope on the top scope. So we shouldn't have a case where top scope stores the class scope, but an inner scope stores the node scope.

> The other question is whether we might inadvertently look up the wrong scope elsewhere, since this fix will now allow us to have two "class" scopes with the same name.

I don't think this can happen either, as we there can only be one class or node scope with that name.

That said, https://tickets.puppetlabs.com/browse/PUP-3995 argues for the opposite behavior. If a node name overlaps with a class name, then it should be an error. I think that would be easy to achieve by adding the check back to `class_set`. Thoughts?

Reply:
Being able to have two scopes with the same name seems like a time bomb. Given that this feature (having a class and node with the same name) has never actually _worked_, it seems reasonable to make it explicitly fail if making it work would introduce this negative side effect.

I don't think node statements are particularly en vogue these days anyway (vs. `include(lookup('classes'))`), so making them _more_ functional probably isn't super important.

Reply:
> it seems reasonable to make it explicitly fail if making it work would introduce this negative side effect... it seems reasonable to make it explicitly fail if making it work would introduce this negative side effect

Yeah that makes sense. One wrinkle is using a regexp does work due to https://tickets.puppetlabs.com/browse/PUP-3167. So the string vs regexp ambiguity is a little odd.

```puppet
node /abc/ { include abc }
```

But like you said, making it more functional isn't super important given other ways of classification.